external_api

LukeP91 opened the pull request.

Commit:
update todo.md

Commit:
add twitter gem. update todo files

Commit:
worker now saves tweets and hashtags to db

Commit:
add sidekiq-cron

Commit:
add scheduler for twiiter_worker

* it will be run every 2 minuts

Commit:
add twitter wrapper

Commit:
add tweet id from twitter

Commit:
apply some refactoring

Commit:
inline one of doubles

simon2k commented on app/models/hash_tag.rb:
```diff
@@ -0,0 +1,4 @@
+class HashTag < ApplicationRecord
```
add missing validation for all models fields

simon2k commented on spec/wrappers/twitter_wrapper_spec.rb:
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+describe TwitterWrapper do
+  describe '#fetch?' do
```
remove question mark

simon2k commented on spec/workers/twitter_worker_spec.rb:
```diff
@@ -0,0 +1,70 @@
+describe TwitterWorker do
+  it 'saves tweets to DB' do
+    last_tweet = double("LastTweet")
+    tweet = { user_name: 'luke_pawlik', message: 'New blog post is up. #Ruby #Rails', hashtags: %w[Ruby Rails], tweet_id: 1 }
```
✂️  remove the `tweet` var from these examples

simon2k commented on app/workers/twitter_worker.rb:
```diff
@@ -0,0 +1,21 @@
+  def perform(hashtags)
+    last_tweet_id = Tweet.last.tweet_id || 0
+    tweets = TwitterWrapper.new.fetch(hashtags, last_tweet_id )
+    save_to_databse(tweets)
```
iterate on the tweets collection returned from `fetch`:

```ruby
TwitterWrapper.new.fetch(hashtags, last_tweet_id).each { |tweet| save(tweet) }
```

simon2k commented on app/wrappers/twitter_wrapper.rb:
```diff
@@ -0,0 +1,18 @@
+class TwitterWrapper
+  def initialize
+    @client = Twitter::REST::Client.new do |config|
+      config.consumer_key        = Rails.application.secrets.consumer_key
```
is that reformatted by rubocop?

simon2k commented on app/wrappers/twitter_wrapper.rb:
```diff
@@ -0,0 +1,18 @@
+  end
+
+  def fetch(hashtags, last_tweet_id)
+    search_query = hashtags.join(' OR ') + ' -rt'
```
use string interpolation instead of string concatenation

simon2k commented on app/wrappers/twitter_wrapper.rb:
```diff
@@ -0,0 +1,18 @@
+  def fetch(hashtags, last_tweet_id)
+    search_query = hashtags.join(' OR ') + ' -rt'
+    @client.search(search_query, count: @tweet_count, since_id: last_tweet_id, result_type: 'recent').map do |tweet|
+      { user_name: tweet.user.name, message: tweet.full_text, hashtags: tweet.hashtags.map(&:text), tweet_id: tweet.id }
```
split to multiple lines

Commit:
apply feedback

Commit:
refactor twitter wrapper tests to use vcr

Commit:
add sources responsible for twitter fetch configuration

Commit:
connect tweets with sources

Commit:
add master worker that spawns new twitter_worker for each source

Commit:
update todo.todo

Commit:
update schedule.yml

Commit:
source name should not be empty

Commit:
source should start with # and should not be break by any whitespace character

simon2k commented on app/policies/source_policy.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  def destroy?
+    user.admin? && (record.id != user.id)
```
do we need to check the id difference?

LukeP91 replied:
I think it should protect admin from deleting himself. And it was one of requirements

simon2k commented on app/workers/twitter_worker.rb:
```diff
@@ -1,21 +1,19 @@
     last_tweet_id = Tweet.last.tweet_id || 0
-    tweets = TwitterWrapper.new.fetch(hashtags, last_tweet_id )
-    save_to_databse(tweets)
+    tweets = TwitterWrapper.new.fetch([@source.name], last_tweet_id).each { |tweet| save(tweet) }
```
pass a name as a string instead of an array with hashtags

simon2k commented on spec/features/admin/twitter/admin_sources_new_spec.rb:
```diff
@@ -0,0 +1,22 @@
+
+describe 'Admin Twitter source index' do
+  context 'user with admin privileges' do
+    scenario 'can see only sources from his organization' do
```
here sources should be already created - at least two: one from user organization and the other one from a different organization

simon2k commented on spec/features/admin/twitter/admin_sources_index_spec.rb:
```diff
@@ -0,0 +1,24 @@
+
+describe 'Admin Twitter source new' do
+  context 'user with admin privileges' do
+    scenario 'can create new source for his organization' do
```
most likely the names are set wrongly for these test cases - should this one be below?

Commit:
apply feedback

Commit:
add .tweets_per_hash_tag

* it returns hash that contains all hashtags from given organization with count of tweets

Commit:
return sorted array with most common words

Commit:
add word unification

simon2k commented on app/models/hash_tag.rb:
```diff
@@ -3,4 +3,12 @@ class HashTag < ApplicationRecord
+
+  scope :in_organization, ->(organization) { HashTagsFromOrganizationQuery.call(organization: organization) }
+
+  def self.tweets_count_per_hashtag(organization)
```
move this to calculation

simon2k commented on app/models/hash_tag.rb:
```diff
@@ -3,4 +3,12 @@ class HashTag < ApplicationRecord
 
   validates :name, presence: true
+
+  scope :in_organization, ->(organization) { HashTagsFromOrganizationQuery.call(organization: organization) }
```
Use: `scope :recenty_activated, RecentlyActivatedUsersQuery`

simon2k commented on app/models/tweet.rb:
```diff
@@ -5,4 +5,39 @@ class Tweet < ApplicationRecord
+
+  scope :in_organization, ->(organization) { joins(:sources).where(sources: { organization_id: organization.id }) }
+
+  def self.most_common_words(organization)
```
calculations :v:

simon2k commented on app/models/tweet.rb:
```diff
@@ -5,4 +5,39 @@ class Tweet < ApplicationRecord
+    words_count
+  end
+
+  def self.unify(word)
```
check out rails class / gem responsible for unifying words

simon2k replied:
add `rails` as a test case - should stay `rails`

simon2k commented on app/models/tweet.rb:
```diff
@@ -5,4 +5,39 @@ class Tweet < ApplicationRecord
+
+  def self.message_words_count(message)
+    words_count = {}
+    message.split(' ').each do |word|
```
make this as an instance method of `Tweet`

simon2k commented on app/models/tweet.rb:
```diff
@@ -5,4 +5,39 @@ class Tweet < ApplicationRecord
+    words_count = {}
+    message.split(' ').each do |word|
+      word = unify(word)
+      if words_count[word].present?
```
group by a word -> get count for each group

simon2k commented on app/models/tweet.rb:
```diff
@@ -5,4 +5,39 @@ class Tweet < ApplicationRecord
+
+  def self.messages_words_count(organization)
+    Tweet.in_organization(organization)
+      .map { |tweet| message_words_count(tweet.message) }
```
concatenate all messages and get stats for the large string

simon2k replied:
calcaulate words longer than 3 chars

simon2k replied:
check if it's possible to do that in the DB

simon2k commented on spec/models/tweet_spec.rb:
```diff
@@ -0,0 +1,25 @@
+          ["fourth", 2],
+          ["ruby", 2],
+          ["third", 1],
+          ["first", 1]
```
if there's the same count, sort alphabetically

simon2k commented on app/wrappers/twitter_wrapper.rb:
```diff
@@ -9,9 +9,13 @@ def initialize
+      since_id: last_tweet_id,
+      result_type: 'recent'
+    ).attrs[:statuses].map do |tweet|
       tweet_data(tweet)
```
add test for the data in batches

Commit:
apply feedback

Commit:
add organization to hashtag

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -0,0 +1,38 @@
+    words_occurrences
+      .sort do |a, b|
+        comp = b[1] <=> a[1]
+        comp.zero? ? a[0] <=> b[0] : comp
```
Ojoj 😈  change `a`, `b` and `comp` to more appropriate names.

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -0,0 +1,38 @@
+  def words_occurrences
+    tweets_messages
+      .map { |word| unify(word) }
+      .group_by { |word| word }
```
`.group_by(&:itself)`

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -0,0 +1,38 @@
+    tweets_messages
+      .map { |word| unify(word) }
+      .group_by { |word| word }
+      .map { |word| [word[0], word[1].size] }
```
use two args in this map: `map { |word, occurrences| [word, occurrences] }`

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -0,0 +1,38 @@
+      .gsub(/[`']/, '')
+      .scan(/\w+/)
+      .first
+      .downcase
```
can we simplify this expression? 🤔

Commit:
add statistics page.

It will be used to display charts with most used words and tweets count per hashtag.
All of data will be scoped to admin's organization

simon2k commented on app/calculations/tweets_count_per_hashtag.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def result
+    HashTag.in_organization(organization).inject({}) do |count, tag|
+      count.merge(tag.name.downcase.to_sym => tag.tweets.count)
```
add a cache counter for tweets - check Rails API doc for that

simon2k commented on app/models/tweets_source.rb:
```diff
@@ -0,0 +1,4 @@
+class TweetsSource < ApplicationRecord
+  belongs_to :tweet
+  belongs_to :source
```
add missing validations for belongs to ✌️

simon2k commented on spec/calculations/most_common_words_spec.rb:
```diff
@@ -0,0 +1,26 @@
+          ["ruby", 2],
+          ["first", 1],
+          ["oreilly", 1],
+          ["tweets", 1]
```
minor: single quotes ✌️

simon2k commented on spec/workers/twitter_worker_spec.rb:
```diff
@@ -0,0 +1,138 @@
+    expect(HashTag.pluck(:name)).to include('ruby', 'rails')
+  end
+
+  it 'associate source with tweet' do
```
`associates source with tweet`

Commit:
major update

* apply feedback
* apply fix to spawner which was passing source object instead of id of source
* currently source id is passed to twitter_worker which finds corresponding source
* add statistics for tweets count per hashtag
* add statistics for most common words in tweets messages

Commit:
update todo.todo

Commit:
update old todo tasks

simon2k commented on app/assets/javascripts/tweets_per_hashtag_chart.js:
```diff
@@ -0,0 +1,43 @@
+$(document).ready(() => {
+  const $container = $('<div>').appendTo($('#tweets_per_hashtag'));
+    const hashtags = $('#tweets_per_hashtag').data('tweetsStats').map((x) => x[0]);
```
indent; rename `x`

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -2,37 +2,45 @@ class MostCommonWords < Patterns::Calculation
   def tweets_messages
-    Tweet.in_organization(organization).map(&:message).map(&:split).flatten
+    Tweet.in_organization(organization)
+      .map(&:message)
```
prefer `pluck` over `map` for attributes

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -2,37 +2,45 @@ class MostCommonWords < Patterns::Calculation
-      .scan(/\w+/)
-      .first
-      .downcase
+    word.gsub(/(['`][sS][\W]*\Z|[\W])/, '').downcase
```
pull out `/(['`][sS][\W]*\Z|[\W])/` to a const in this class - usually we pull regexps and numbers to a const at the top

simon2k commented on app/calculations/tweets_count_per_hashtag.rb:
```diff
@@ -2,12 +2,19 @@ class TweetsCountPerHashtag < Patterns::Calculation
     options.fetch(:organization)
   end
+
+  def tweets_count
```
rename to `tag_names_stats` or `hashtags_stats`

simon2k commented on app/wrappers/twitter_wrapper.rb:
```diff
@@ -9,6 +9,7 @@ def initialize
     @tweet_count = 20
   end
 
+
```
✂️

simon2k commented on app/assets/javascripts/most_common_words_chart.js:
```diff
@@ -0,0 +1,45 @@
+$(document).ready(() => {
+  const $container = $('<div>').appendTo($('#most_common_words'));
+  const words = $('#most_common_words').data('wordsStats').map((x) => x[0]);
+  const words_counts = $('#most_common_words').data('wordsStats').map((x) => x[1]);
```
camelCase for JS

simon2k replied:
also, in JS, use trailing comma at the end of line in objects / arrays

Commit:
apply feedback

Commit:
add show and index for hash tags

Admin can now see list of all hash tags within organization.
For each hashtag there is possibility to see last 20 fetched tweets.

Commit:
hash tag show page shows now number of tweets with given hash tag

Commit:
hashtags are saved with # at the beginning

Commit:
MostCommonWords accepts list of tweets as param

* if such list is provided words will be counted using only those tweets

Commit:
save tweet to database together with tweet creation date.

show tweet creation date on hashtag view

Commit:
remove turbolinks

Commit:
format date of tweet creation

simon2k commented on app/calculations/most_common_words.rb:
```diff
@@ -13,6 +13,11 @@ def limit
+    options.fetch(:tweets, Tweet.in_organization(organization))
+  end
+
+
```
✂️

simon2k commented on app/calculations/tweets_count_per_hashtag.rb:
```diff
@@ -2,19 +2,27 @@ class TweetsCountPerHashtag < Patterns::Calculation
   def tweets_count
     HashTag.in_organization(organization).inject([]) do |count, tag|
-      count << [tag.name, tag.tweets.count]
+      count << [tag.name, tag.tweets_count]
```
rename `count` to `stat` and pass there object: `{name: tag.name, count: tag.tweets_count}` - it'll be better for understanding that in the front-end code, so that instead of using cryptic indexes: `[0]` or `[1]`, we'll be able to call e.g.: `stats.name`

Commit:
apply feedback

Commit:
check for uniqueness of tweet_id within organization

Commit:
apply feedback
